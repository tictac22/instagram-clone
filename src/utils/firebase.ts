// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app"

import {
	createUserWithEmailAndPassword,
	deleteUser,
	FacebookAuthProvider,
	getAuth,
	signInWithPopup
} from "firebase/auth"
import {
	collection,
	doc,
	getDocs,
	getFirestore,
	query,
	setDoc,
	where
} from "firebase/firestore"
const firebaseConfig = {
	apiKey: import.meta.env.VITE_apiKey,
	authDomain: import.meta.env.VITE_authDomain,
	projectId: import.meta.env.VITE_projectId,
	storageBucket: import.meta.env.VITE_storageBucket,
	messagingSenderId: import.meta.env.VITE_messagingSenderId,
	appId: import.meta.env.VITE_appId,
	measurementId: import.meta.env.VITE_measurementId
}

// Initialize Firebase
const app = initializeApp(firebaseConfig)
export const db = getFirestore(app)
export const auth = getAuth()

type SignUp = {
	email: string
	password: string
	username: string
	fullName: string
}

export const signInWithFacebook = async () => {
	try {
		const provider = new FacebookAuthProvider()
		const result = await signInWithPopup(auth, provider)

		const usersCollectionRef = doc(db, "users", result.user.uid)
		await setDoc(usersCollectionRef, {
			userId: result.user.uid
		})
	} catch (error) {
		console.log(error)
		throw new CustomError(ErrorMessages.email, "email")
	}
}
export const signUp = async ({
	email,
	password,
	username,
	fullName
}: SignUp) => {
	try {
		const registration = await createUserWithEmailAndPassword(
			auth,
			email,
			password
		)
		const isTaken = await isNameTaken(username)
		if (isTaken) {
			await deleteUser(registration.user)
			throw new CustomError(ErrorMessages.username, "username")
		}
		const usersCollectionRef = doc(db, "users", registration.user.uid)
		await setDoc(usersCollectionRef, {
			userId: registration.user.uid,
			username,
			fullName
		})
	} catch (error) {
		if (error instanceof CustomError) {
			throw error
		}
		throw new CustomError(ErrorMessages.email, "email")
	}
}
const isNameTaken = async (username: string) => {
	try {
		const usersCollectionRef = collection(db, "users")
		const q = query(usersCollectionRef, where("username", "==", username))

		const querySnapshot = await getDocs(q)

		const users = []
		querySnapshot.forEach((item) => {
			users.push({ ...item.data(), id: item.id })
		})
		if (users.length >= 1) return true
		return false
	} catch (error) {
		console.error(error)
	}
}

class CustomError extends Error {
	status: number
	data: object
	constructor(message: string, field: string) {
		super(message)
		this.status = 404
		this.data = {
			[field]: [message]
		}
	}
}

const enum ErrorMessages {
	email = "Email is already in use",
	username = "Username is already in use"
}
