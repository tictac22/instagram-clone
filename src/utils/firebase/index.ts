export * from "./auth"
export * from "./config"
export * from "./post"
export * from "./storage"
