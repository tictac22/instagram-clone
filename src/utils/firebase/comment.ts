import {
	addDoc,
	collection,
	doc,
	increment,
	Timestamp,
	updateDoc
} from "firebase/firestore"
import { db } from "./config"

type CommentData = {
	photoUrl: string
	userName: string
	id: string
	text: string
}
export const createCommentPost = async (data: CommentData) => {
	const postCollection = doc(db, "posts", data.id)
	await updateDoc(postCollection, {
		comments: increment(1)
	})

	const collections = collection(db, "posts", data.id, "comments")
	const response = await addDoc(collections, {
		userName: data.userName,
		photoUrl: data.photoUrl ?? "",
		text: data.text,
		createdAt: Timestamp.fromDate(new Date())
	})
	return response
}
