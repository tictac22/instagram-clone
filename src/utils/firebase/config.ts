// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app"

import { getAuth } from "firebase/auth"
import { getFirestore } from "firebase/firestore"
const firebaseConfig = {
	apiKey: import.meta.env.VITE_apiKey,
	authDomain: import.meta.env.VITE_authDomain,
	projectId: import.meta.env.VITE_projectId,
	storageBucket: import.meta.env.VITE_storageBucket,
	messagingSenderId: import.meta.env.VITE_messagingSenderId,
	appId: import.meta.env.VITE_appId,
	measurementId: import.meta.env.VITE_measurementId
}

// Initialize Firebase
const app = initializeApp(firebaseConfig)
export const db = getFirestore(app)
export const auth = getAuth()

export class CustomError extends Error {
	status: number
	data: object
	constructor(message: string, field: string) {
		super(message)
		this.status = 404
		this.data = {
			[field]: [message]
		}
	}
}

export const enum ErrorMessages {
	email = "Email is already in use",
	username = "Username is already in use",
	password = "password is incorrect"
}

export type SignUp = {
	email: string
	password: string
	username: string
	fullName: string
}

export type Login = {
	email: string
	password: string
}
