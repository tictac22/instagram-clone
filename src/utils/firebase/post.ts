import {
	addDoc,
	collection,
	doc,
	documentId,
	getDoc,
	getDocs,
	increment,
	orderBy,
	query,
	Timestamp,
	updateDoc,
	where
} from "firebase/firestore"
import { Author, Post } from "../types"
import { db } from "./config"

export const getUserHomePosts = async (subscribedIds: string[]) => {
	if (!subscribedIds || subscribedIds.length < 0) return
	const postCollection = collection(db, "posts")
	const q = query(
		postCollection,
		orderBy("createdAt", "desc"),
		where("uid", "in", subscribedIds)
	)

	const querySnapshot = await getDocs(q)
	const posts: Post[] = []
	//@ts-ignore
	querySnapshot.forEach(doc => posts.push({ id: doc.id, ...doc.data() }))

	const userCollection = collection(db, "users")
	const notRepeatedUserUids = [...new Set(posts.map(item => item.uid))]
	const q2 = query(
		userCollection,
		where(documentId(), "in", notRepeatedUserUids)
	)
	const userSnapShot = await getDocs(q2)
	const users: Author[] = []

	//@ts-ignore
	userSnapShot.forEach(doc => users.push({ id: doc.id, ...doc.data() }))

	posts.forEach(
		//@ts-ignore
		item => (item.author = users.find(user => user.id === item.uid))
	)

	return posts
}

export const getPost = async (id: string) => {
	const postRef = doc(db, "posts", id)

	const postRequest = await getDoc(postRef)
	const post = postRequest.data() as Post

	const userRef = doc(db, "users", post.uid)

	const userRequest = await getDoc(userRef)
	const user = userRequest.data() as Author

	return {
		post: {
			...post,
			id: postRequest.id
		},
		user: {
			...user,
			uid: userRequest.id
		}
	}
}

interface ICreatePost {
	images: string[]
	creatorUid: string
	text: string
}

export const createPost = async (postData: ICreatePost) => {
	const collections = collection(db, "posts")
	const data = await addDoc(collections, {
		...postData,
		createdAt: Timestamp.fromDate(new Date()),
		likes: 0,
		comments: 0
	})
	return data
}
export const likePost = async (
	id: string,
	postId: string,
	likes: string[],
	liked: boolean
) => {
	const post = doc(db, "posts", postId)
	if (liked) {
		await updateDoc(post, {
			likes: increment(-1)
		})
	} else {
		await updateDoc(post, {
			likes: increment(1)
		})
	}
	const userRef = doc(db, "users", id)
	updateDoc(userRef, {
		likes: [...likes]
	})
}
