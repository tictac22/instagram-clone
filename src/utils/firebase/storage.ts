import { getDownloadURL, getStorage, ref, uploadBytes } from "firebase/storage"
import { v4 as uuidv4 } from "uuid"
import { app } from "./config"

// Create a root reference
const storage = getStorage(app)
const metadata = { contentType: "image/jpeg" }

export const saveFiles = async (imgUrl: Blob[]) => {
	const images: string[] = []

	for (let index = 0; index < imgUrl.length; index++) {
		const element = imgUrl[index]
		const file = await blobToFile(element)
		const imageRef = ref(storage, "post/" + uuidv4())

		await uploadBytes(imageRef, file, metadata)
		const image = await getDownloadURL(imageRef)
		images.push(image)
	}

	return images
}

const blobToFile = async (image: Blob) => {
	//@ts-ignore
	const result = await fetch(image)
	const blobFile = await result.blob()
	return new File([blobFile], uuidv4(), { type: blobFile.type })
}
