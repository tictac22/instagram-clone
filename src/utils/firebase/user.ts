import {
	collection,
	DocumentData,
	getDocs,
	query,
	where
} from "firebase/firestore"
import { db } from "./config"

export const suggestUser = async (username: string) => {
	const parsedUserName = username.toLowerCase().substring(1)
	const q = query(
		collection(db, "users"),
		where("userName", ">=", parsedUserName),
		where("userName", "<=", parsedUserName + "\uf8ff")
	)
	const users = await getDocs(q)
	const data: DocumentData[] = []
	users.forEach(doc => {
		data.push(doc.data())
	})
	return data
}
