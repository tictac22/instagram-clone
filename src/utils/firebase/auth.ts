import {
	createUserWithEmailAndPassword,
	deleteUser,
	FacebookAuthProvider,
	signInWithEmailAndPassword,
	signInWithPopup,
	signOut
} from "firebase/auth"
import {
	collection,
	doc,
	getDoc,
	getDocs,
	query,
	setDoc,
	where
} from "firebase/firestore"
import { auth, CustomError, db, ErrorMessages, Login, SignUp } from "./config"
import { savePhotoUrl } from "./storage"

export const logOut = async () => {
	return await signOut(auth)
}

export const signInWithFacebook = async () => {
	try {
		const provider = new FacebookAuthProvider()
		await signInWithPopup(auth, provider)
	} catch (error) {
		throw new CustomError(ErrorMessages.email, "email")
	}
}
export const setSignUpFacebookData = async ({
	uid,
	fullName,
	userName,
	photoUrl
}: {
	uid: string
	fullName: string
	userName: string
	photoUrl: string
}) => {
	const isTaken = await isNameTaken(userName)
	if (isTaken) {
		throw new CustomError(ErrorMessages.username, "userName")
	}
	const usersCollectionRef = doc(db, "users", uid)

	const user = await setDoc(usersCollectionRef, {
		fullName,
		userName,
		photoUrl,
		likes: [],
		subscribed: [],
		followers: 0
	})
	return user
}
export const signUp = async ({
	email,
	password,
	userName,
	fullName,
	photoUrl = ""
}: SignUp) => {
	try {
		const registration = await createUserWithEmailAndPassword(
			auth,
			email,
			password
		)
		const isTaken = await isNameTaken(userName)
		if (isTaken) {
			await deleteUser(registration.user)
			throw new CustomError(ErrorMessages.username, "username")
		}
		const usersCollectionRef = doc(db, "users", registration.user.uid)
		const image = photoUrl.includes("blob")
			? await savePhotoUrl(photoUrl)
			: ""
		await setDoc(usersCollectionRef, {
			userId: registration.user.uid,
			userName,
			fullName,
			likes: [],
			photoUrl: image,
			subscribed: [],
			followers: 0
		})
	} catch (error) {
		if (error instanceof CustomError) {
			throw error
		}
		throw new CustomError(ErrorMessages.email, "email")
	}
}

export const logIn = async ({ email, password }: Login) => {
	try {
		const user = await signInWithEmailAndPassword(auth, email, password)
		return user
	} catch (error) {
		//@ts-ignore
		const message = error.message
		if (message === firebaseError.email) {
			throw new CustomError(ErrorMessages.email, "email")
		}
		throw new CustomError(ErrorMessages.password, "email")
	}
}

async function isNameTaken(username: string) {
	try {
		const usersCollectionRef = collection(db, "users")
		const q = query(usersCollectionRef, where("username", "==", username))

		const querySnapshot = await getDocs(q)

		const users = []
		querySnapshot.forEach(item => {
			users.push({ ...item.data(), id: item.id })
		})
		if (users.length >= 1) return true
		return false
	} catch (error) {
		//eslint-disable-next-line
		console.error(error)
	}
}

const enum firebaseError {
	username = "Username is already in use",
	email = "Firebase: Error (auth/user-not-found).",
	password = "Firebase: Error (auth/wrong-password)."
}

export const getUser = async (userId: string) => {
	const usersCollectionRef = doc(db, "users", userId)
	const user = await getDoc(usersCollectionRef)
	return user.data() ?? false
}
