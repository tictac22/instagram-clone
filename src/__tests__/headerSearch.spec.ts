import Search from "@/components/header/search.vue"
import { suggestUser } from "@/utils/firebase"
import { mount, VueWrapper } from "@vue/test-utils"
import { beforeEach, describe, expect, it, vi } from "vitest"
import { nextTick } from "vue"

import { router } from "./helpers/renderWithRouter"

vi.mock("@/utils/firebase")
const fakeUser = [
	{
		uid: "1",
		fullName: "333",
		userName: "333",
		postsNumber: 0,
		followers: 0,
		photoUrl: ""
	}
]
const suggest = vi.mocked(suggestUser).mockResolvedValue(fakeUser)

describe("headSearch component", () => {
	let wrapper: VueWrapper
	beforeEach(async () => {
		wrapper = mount(Search, {
			global: {
				plugins: [router],
				stubs: ["FontAwesomeIcon"],
				directives: {
					clickOutside: () => {}
				}
			},
			props: {
				searchInput: "",
				"onUpdate:searchInput": (e: string) =>
					wrapper.setProps({ searchInput: e })
			}
		})

		await wrapper.find("input").setValue("test")
		await new Promise(r => setTimeout(r, 2000))

		await suggest("elon musk")
		await nextTick()
	}, 4000)
	it("mounts successfuly", () => {
		expect(wrapper.text()).toContain("Recent")
	})
	it("fetchs data and render them", () => {
		expect(wrapper.find('[data-test="1"]')).toBeTruthy()
	})
	it("adds searched user to state", async () => {
		const spy = vi.spyOn(Storage.prototype, "setItem")

		wrapper.find('[data-test="1"]').trigger("click")

		expect(spy).toHaveBeenCalledWith(
			"searchedUsers",
			JSON.stringify(fakeUser)
		)
		expect(wrapper.element).toMatchSnapshot()
	})
	it("deletes history", async () => {
		const spy = vi.spyOn(Storage.prototype, "removeItem")

		await wrapper.find('[data-test="1"]').trigger("click")
		await wrapper.find("#deleteButton").trigger("click")

		expect(spy).toHaveBeenCalledWith("searchedUsers")
	})
})
