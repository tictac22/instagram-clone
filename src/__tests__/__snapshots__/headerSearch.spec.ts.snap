// Vitest Snapshot v1

exports[`headSearch component > adds searched user to state 1`] = `
<div
  class="relative h-9"
  searchinput=""
>
  <div
    class="pointer-events-none absolute left-2 top-1/2 -translate-y-1/2 cursor-pointer text-gray-500"
  >
    <font-awesome-icon-stub
      icon="fa-solid fa-magnifying-glass"
    />
  </div>
  <input
    class="h-full min-w-[268px] rounded-lg bg-[#efefef] px-3 pl-7 outline-none"
    placeholder="Search"
  />
  <!--v-if-->
  <div
    class="absolute top-[50px] z-10 h-[375px] w-[375px] -translate-x-12 rounded-lg bg-white drop-shadow transition-all pointer-events-none invisible -translate-y-3 opacity-0"
  >
    <div
      class="triangle -top-[13px] left-1/2 -translate-x-1/2"
    />
    <div
      class="relative flex h-full flex-col overflow-y-scroll"
    >
      <div
        class="flex items-center justify-between p-3"
      >
        <h4
          class="font-bold"
        >
          Recent
        </h4>
        <p
          class="cursor-pointer text-blue-300"
          id="deleteButton"
        >
           clear all 
        </p>
      </div>
      
      <a
        class="flex w-full cursor-pointer items-center pl-3 hover:bg-[rgb(250,250,250)]"
        data-test="1"
        href="/1"
      >
        <img
          class="h-8 w-8 rounded-full"
          src=""
        />
        <div
          class="ml-3"
        >
          <p
            class="font-semibold"
          >
            333
          </p>
          <p
            class="font-medium text-gray-200"
          >
            333
          </p>
        </div>
      </a>
      
      <div
        class="flex flex-auto items-center self-center text-gray-400"
      >
        <!--v-if-->
        <!--v-if-->
      </div>
    </div>
  </div>
</div>
`;
